Feat/display student list

Shiva30jun opened the pull request:
# Summary

This PR implements the `StudentList` component to display a paginated student table with support for dependency injection and empty state handling.

# Changes Included

- **Display student data in table:**
  - Name, Email, Role, Section, Group columns
  - Email button 
- **Empty state:**
  - Renders "No students found." message if list is empty

- **Tests:**
  - Verify rendering of student rows
  - Check email button is available
  - Confirm pagination summary
  - Ensure empty state displays correctly

# Testing

  - Render student data correctly
  - Display empty state message
  - Show correct pagination summary for populated and empty data
  
  
  Related GitHub Issues: 
  https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/596
  https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/597

shiva-amanaganti committed:
feat(StudentList): add studentGetter prop and implement table row rendering

shiva-amanaganti committed:
test(StudentList): add UI test for rendering student row with all columns

shiva-amanaganti committed:
test(StudentList): verify email link href is rendered correctly

shiva-amanaganti committed:
test(StudentList): verify pagination summary displays correct range

shiva-amanaganti committed:
test(StudentList): verify empty state message renders when no students

shiva-amanaganti committed:
test(StudentList): verify pagination summary when no students

shiva-amanaganti committed:
refactor: replace <Link> with <button> for email action icon

shiva-amanaganti committed:
test: updated the test case to show the email button

achow-conestogac commented:
a couple of problems here... tests do not run.

also, where is the usage?

shiva-amanaganti committed:
updated the failing test case

shiva-amanaganti committed:
feat: add /students route to render StudentList component

shiva-amanaganti committed:
chore: ensured that the branch feat/display_student_list is upto date with main

shiva-amanaganti committed:
updated the app.tsx with /students route to render StudentList component

shiva-amanaganti committed:
Merge branch 'feat/display_student_list' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/display_student_list

Shiva30jun commented:
Hello professor @achow-conestogac 
Regarding the usage, the routing is now set up so that navigating to /students will display the StudentList component.

shiva-amanaganti committed:
Refactor: moved the route of Student List

shiva-amanaganti committed:
Merge branch 'main' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/display_student_list

shiva-amanaganti committed:
Updated the app.tsx

achow-conestogac commented:
hello @shiva-amanaganti `npm run dev` is error out. 

please fix that.

shiva-amanaganti committed:
Updated the index file

shiva-amanaganti committed:
Updated the imports

shiva-amanaganti committed:
feat: add the index file for student list

shiva-amanaganti committed:
Added the outlet in dashboard so that the student table data renders in the route enrollment/students

Shiva30jun commented:
Hello Professor,
All the necessary changes are now made.'npm run dev' should not give any errors now.
Able to render in "/enrollment/students". 
Could you please review it and provide necessary feedback.

achow-conestogac commented on src/Enrollment/Dashboard/Dashboard.tsx:
```diff
@@ -21,6 +21,7 @@ const Dashboard: React.FC = () => {
           home page
         </Link>{" "}
       </p>
+      <Outlet />
```
What is the purpose of this addition?

Shiva30jun replied:
removed the <Outlet />

achow-conestogac commented on src/App.tsx:
```diff
@@ -4,11 +4,14 @@ import logo from './assets/logo.png';
     <BrowserRouter>
       <Routes>
-        <Route path="/enrollment" element={<EnrollmentDashboard />} />
+        <Route path="/enrollment" element={<EnrollmentDashboard />} >
+          <Route path="students" element={<StudentList />} />
+        </Route>
```
As we clean up this pull request, I am left wondering why we need an `/enrollment/students` route.

why can't we simply put this on the `src/Enrollment/Dashboard` page?

on `src/Enrollment/Student/Pages/StudentList/StudentList.tsx` your type definition involves using a `studentGetter` function.

but if you simply allow the `EnrollmentDashboard` to handle getting the data, then you can use it without having to get it yourself again.

I feel strongly that introducing a new `route` is not the best way of moving forward.  Please integrate the `StudentList` into the `Dashboard` directly.

Shiva30jun replied:
As per the feedback used the StudentList into the Dashboard directly.

achow-conestogac commented on src/Enrollment/Student/Pages/StudentList/StudentList.tsx:
```diff
@@ -15,7 +20,7 @@ const StudentList: React.FC = () => {
   const pageSize = 10;
 
-  const { data: students, total, totalPages } = getStudents(page, pageSize);
+  const { data: students, total, totalPages } = getStudents(studentGetter, page, pageSize);
```
this function signature makes no sense.

why does the `getStudent` function require a function that gets students.

achow-conestogac replied:
no. this makes no sense. 

the `getStudents` function is already something that does this job.  nesting functionality like this is not helpful.

Shiva30jun replied:
Refactored the code removing the nesting functionality

shiva-amanaganti committed:
Refactor: used the local storage to show the student data

shiva-amanaganti committed:
Refactor: Updated the test case as per the update in the StudentList

Shiva30jun commented:
Hello Professor @achow-conestogac 
Updated the code as for the feedbacks.
Can you please review them?

shiva-amanaganti committed:
Implement pagination logic in paginateStudents.ts for student data from localStorage

shiva-amanaganti committed:
Update tests for paginateStudents.ts to reflect pagination changes

shiva-amanaganti committed:
Refactor and update tests for StudentList.tsx to eliminate unnecessary nesting

shiva-amanaganti committed:
Added test in the Dashboard.test.tsx as per the update

Shiva30jun commented on src/Enrollment/Dashboard/Dashboard.tsx:
Added the test in the Dashboard.test.tsx

Shiva30jun commented:
Hello Professor
@achow-conestogac 
Can you please review my updated code?

Thankyou

Shiva30jun commented:
Hello Professor
@achow-conestogac
Could you please review my updated code?

Thank you

achow-conestogac commented on src/Enrollment/Student/studentData/paginateStudents.ts:
```diff
@@ -1,12 +1,15 @@
 ): { data: Student[]; total: number; totalPages: number } {
-  const students = getAllStudents(studentGetter);
-  if (!students || !Array.isArray(students)) return { data: [], total: 0, totalPages: 0 };
+  const studentDataFromCache = localStorage.getItem("students_list_key");
```
you need `LocalStorageService`

Shiva30jun replied:
Hello Professor @achow-conestogac,
I have now integrated the LocalStorageService into paginateStudents.ts for fetching student data. This ensures all storage operations are handled through our centralized service, improving consistency, type safety, and maintainability of the codebase.

Shiva30jun replied:
Hello professor @achow-conestogac 
Can you please review my updated code?
Thank you.

Shiva30jun replied:
Hello professor @achow-conestogac
Could you please review my updated code?
Thank you.

shiva-amanaganti committed:
feat(pagination): integrate LocalStorage service for student data retrieval

Implemented LocalStorage wrapper in getStudents for consistent, type-safe, and centralized storage handling.

shiva-amanaganti committed:
Updated the import path

shiva-amanaganti committed:
test(pagination): update getStudents tests to use LocalStorage service

Refactored unit tests for getStudents to use the LocalStorage service for data setup, aligning test behavior with production storage handling

shiva-amanaganti committed:
chore: Removed the unnecessary comments
